contract_comm/evm.go

```diff
@@ -18,139 +18,76 @@ package contract_comm
 
 import (
 	"math/big"
-	"reflect"
-	"time"
 
 	"github.com/celo-org/celo-blockchain/accounts/abi"
 	"github.com/celo-org/celo-blockchain/common"
-	"github.com/celo-org/celo-blockchain/common/hexutil"
 	"github.com/celo-org/celo-blockchain/contract_comm/errors"
 	"github.com/celo-org/celo-blockchain/contracts"
 	"github.com/celo-org/celo-blockchain/core/types"
 	"github.com/celo-org/celo-blockchain/core/vm"
-	"github.com/celo-org/celo-blockchain/core/vm/vmcontext"
 	"github.com/celo-org/celo-blockchain/log"
-	"github.com/celo-org/celo-blockchain/metrics"
 )
 
 var (
-	internalEvmHandlerSingleton *InternalEVMHandler
+	evmCallerFactory vm.EVMCallerFactory
 )
 
-// An EVM handler to make calls to smart contracts from within geth
-type InternalEVMHandler struct {
-	chain vm.ChainContext
-}
-
-func MakeStaticCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state vm.StateDB) (uint64, error) {
-	return makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, nil, header, state, true)
-}
-
-func MakeCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state vm.StateDB, finaliseState bool) (uint64, error) {
-	gasLeft, err := makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, value, header, state, false)
-	if err == nil && finaliseState {
-		state.Finalise(true)
+func SetEVMCallerFactory(_evmCallerFactory vm.EVMCallerFactory) {
+	if evmCallerFactory == nil {
```

Comment:
Feels funny for it to no-op silently if it's called again with a different factory.  I can think of two alternatives: (a) returning an error or panicking if it's called again, or (b) overwriting.

I guess we have no need for (b) yet, but I could imagine it maybe coming in handy during tests.  And, at first sight, I don't see a big problem with overwriting.

Reply:
I'd be in favour of panicking, this is clearly not something that should be called more than once in production code. There is another problem here though which is that this lacks synchronisation and there doesn't seem to be any guarantee that this would only be called from a single go-routine.

Reply:
can't panic, since tests actually do override this. Acutally i don't trust tests too much, since maybe we have conflicting singleton on them :(

Luckliy next PR actually removes this singleton